### ISSUE TITLE
`delphix` plugin bundles proprietary dependency

### ISSUE THREAD

**Issue by daniel-beck** [jenkins-infra/helpdesk#3989]
### Service(s)

Update center

### Summary

https://plugins.jenkins.io/delphix/ bundles https://github.com/jenkinsci/delphix-plugin/blob/f650195fbfe969816780e9eebedb4f642c0a273f/pom.xml#L97-L101 which is a proprietary dependency. This was first added in 3.0.0 which was basically a rewrite of the plugin, see https://github.com/jenkinsci/delphix-plugin/commit/d1d6b8137cd2d98e8a337a890f3e34d319b3c6cd and https://github.com/jenkins-infra/update-center2/pull/725.

While the plugin itself claims to be licensed Apache 2 per https://github.com/jenkinsci/delphix-plugin/blob/master/LICENSE, we require that plugins are open source "all the way down".

Given the terms of the Delphix license and the fact that the Jenkins project is distributing the plugin with the proprietary dependency, it appears we might even be in violation of their license.

## References
https://www.jenkins.io/project/governance/#3rd-party-library-licenses-in-the-plugins
https://www.jenkins.io/doc/developer/publishing/preparation/#license

> All plugins distributed by the Jenkins project need to be free and open source software. This applies to the plugin source code as well as all its dependencies. Make sure to specify the license both in the pom.xml file, as well as a LICENSE file in your repository.

https://download.delphix.com/file/12099/delphix-developer-license-1_1.txt

## Next Actions

* Reach out to maintainers and give them a bit of time to fix this (CCed below)
* Suspend plugin distribution through update sites / plugins.jenkins.io
* Delete artifacts from Artifactory (IIUC, we're in violation of their license terms by distributing the plugin)

CC @nick-mathison @ankit-patil-hubs as maintainers. @NotMyFault @uhafner @MarkEWaite @basil as board members.

**Comment by nick-mathison** [comment]
Hi @daniel-beck, thanks for reporting this issue. We'll be reviewing it further this week. 

For our complete understanding of the requirements, is the expectation that we publish the source code of this Jar?
https://mvnrepository.com/artifact/com.delphix/dct-api/8.0.1-sdk

Is that accurate?

**Comment by MarkEWaite** [comment]
> For our complete understanding of the requirements, is the expectation that we publish the source code of this Jar? https://mvnrepository.com/artifact/com.delphix/dct-api/8.0.1-sdk
> 
> Is that accurate?

The [governance document](https://www.jenkins.io/project/governance/#license) says the plugin (and thus its dependencies, including the jar file) must be licensed with an [OSI compliant license](https://opensource.org/licenses/).  Publishing the source code of the jar is desirable, but an OSI compliant license is required.  The license of that jar file needs to be changed to an OSI compliant license or the jar file needs to be removed from the Jenkins plugin.

Does that address your question @nick-mathison ?

**Comment by ankit-patil-hubs** [comment]
Hi @MarkEWaite  @daniel-beck , we are working on the resolution of the license for the proprietary dependency.

**Comment by daniel-beck** [comment]
An alternative solution would be to remove the bundled jar from the plugin. Releases before 3.0 did not have it IIRC. (Then we would only suspend distribution of the releases with the offending jar, but new releases could remain.

**Comment by ankit-patil-hubs** [comment]
@daniel-beck Indeed, versions prior to 3.0 did not include this JAR file. However, this particular JAR file is essential for our plugin, and removing it would result in the complete cessation of the plugin’s functionality.

**Comment by daniel-beck** [comment]
Right, some rewriting would be needed. Just mentioning a potential alternative approach if your lawyers tell you "no" 😃

**Comment by ankit-patil-hubs** [comment]
@daniel-beck We're considering creating a new JAR file with an Apache License. Instead of publishing it to Maven Central, we'd keep it local within a new `lib` directory in the Jenkins CI Delphix plugin repository and use it directly from there. Do you think this approach would be feasible? Basically we want to bundle this jar as part of the plugin itself

**Comment by MarkEWaite** [comment]
@ankit-patil-hubs so long as the source code that is included in the new lib directory is licensed with an OSI approved open source license, that would be acceptable for the licensing requirement.  If the source code that is placed in the lib directory is licensed with a proprietary license, then that won't resolve the issue.

**Comment by ankit-patil-hubs** [comment]
@MarkEWaite Got it, Thanks. We are working on resolving this and will release a new version with the fix ASAP. But from a timeline perspective we wanted to understand if we release a new version in the next 4-5 days can we avoid the suspension?

**Comment by daniel-beck** [comment]
> 4-5 days can we avoid the suspension?

That should be fine.

**Comment by ankit-patil-hubs** [comment]
@daniel-beck @MarkEWaite Just an FYI, we have published the dependency again with an open-source license, if all goes well the Jenkins plugin with changes should be out by tomorrow.

**Comment by ankit-patil-hubs** [comment]
We have released a new version of the plugin with the required changes 
https://github.com/jenkinsci/delphix-plugin/releases/tag/delphix-3.2.0

**Comment by MarkEWaite** [comment]
Thanks very much @ankit-patil-hubs .  I installed Delphix plugin 3.1.1 and confirmed it has a component with a proprietary license then upgraded to Delphix plugin 3.2.0 and confirmed that all its components are using open source licenses.

I've submitted a pull request to suspend distribution of the earlier versions. Refer to:

* https://github.com/jenkins-infra/update-center2/pull/768

@daniel-beck should we wait some period of time before merging that pull request so that users see the new release in their Jenkins plugin manager ?

**Comment by daniel-beck** [comment]
> should we wait some period of time before merging that pull request so that users see the new release in their Jenkins plugin manager ?

Only saw this message after merging, so there's a risk now of folks within the next ~20 hrs trying to install 3.1.1. Given the low number of total installs I would say the risk is minimal, and it's resolved by re-requesting updated metadata, a common troubleshooting step, so don't think we need to revert.

----

FTR we need to keep this issue open to track deletion of the unlicensed older releases.

**Comment by MarkEWaite** [comment]
The older releases that need to be deleted from Jenkins Artifactory are:

* https://repo.jenkins-ci.org/ui/repos/tree/General/releases/org/jenkins-ci/plugins/delphix/3.0.1
* https://repo.jenkins-ci.org/ui/repos/tree/General/releases/org/jenkins-ci/plugins/delphix/3.0.2
* https://repo.jenkins-ci.org/ui/repos/tree/General/releases/org/jenkins-ci/plugins/delphix/3.0.3
* https://repo.jenkins-ci.org/ui/repos/tree/General/releases/org/jenkins-ci/plugins/delphix/3.1.0
* https://repo.jenkins-ci.org/ui/repos/tree/General/releases/org/jenkins-ci/plugins/delphix/3.1.1

That will need to be done by an Artifactory administrator.  @daniel-beck and @dduportal are both Artifactory administrators as far as I know.

**Comment by dduportal** [comment]
> The older releases that need to be deleted from Jenkins Artifactory are:
> 
>     * https://repo.jenkins-ci.org/ui/repos/tree/General/releases/org/jenkins-ci/plugins/delphix/3.0.1
> 
>     * https://repo.jenkins-ci.org/ui/repos/tree/General/releases/org/jenkins-ci/plugins/delphix/3.0.2
> 
>     * https://repo.jenkins-ci.org/ui/repos/tree/General/releases/org/jenkins-ci/plugins/delphix/3.0.3
> 
>     * https://repo.jenkins-ci.org/ui/repos/tree/General/releases/org/jenkins-ci/plugins/delphix/3.1.0
> 
>     * https://repo.jenkins-ci.org/ui/repos/tree/General/releases/org/jenkins-ci/plugins/delphix/3.1.1
> 
> 
> That will need to be done by an Artifactory administrator. @daniel-beck and @dduportal are both Artifactory administrators as far as I know.

I'm taking the action (as I also have other artifacts to remove/move in https://github.com/jenkins-infra/helpdesk/issues/4021)

**Comment by dduportal** [comment]
Done. The `maven-metdata.xml` is updated as well as the tree.
Note there are two 3.x SNAPSHOT artifacts (see https://repo.jenkins-ci.org/artifactory/public/org/jenkins-ci/plugins/delphix/): should we remove them as well?

**Comment by dduportal** [comment]
Update:

- Moved the artifacts from repository `archives` to `delphix-helpdesk-3989` (private repository only available by admins: https://repo.jenkins-ci.org/artifactory/delphix-helpdesk-3989/) as per a discussion with @daniel-beck to continue the same process of archiving as he use to do on former request of this kind: we never remove artifacts, we only move them.
- Moved all the `3.x` snapshot artifacts (`3.0.0-SNAPSHOT` and `3.1.1-SNAPSHOT`)
- Removed the (now empty) private `archives` repository (its name is too generic)

https://repo.jenkins-ci.org/artifactory/public/org/jenkins-ci/plugins/delphix/ now only show the 2.x and the 3.2.0 versions.

Handing out to you folks.

**Comment by MarkEWaite** [comment]
Artifactory looks right to me.  It shows the releases before 3.0.0 and the 3.2.0 release.

* https://repo.jenkins-ci.org/ui/repos/tree/General/releases/org/jenkins-ci/plugins/delphix

Updates.jenkins.io looks right to me as well.  It should the 3.2.0 release and does not show the 3.0.x or 3.1.x releases.

* https://updates.jenkins.io/download/plugins/delphix/

**Comment by daniel-beck** [comment]
Anything we're missing or is this done now?

**Comment by MarkEWaite** [comment]
> Anything we're missing or is this done now?

I don't know of anything further that needs to be done.  I believe we are done.
